### ISSUE TITLE
cla-bot not reacting on pull request

### ISSUE THREAD

**Issue by maoo** [finos/cla-bot#23]
I've enabled cla-bot on https://github.com/symphonyoss account (only for 1 project, [https://github.com/symphonyoss/clabot-test](https://github.com/symphonyoss/clabot-test)) and defined https://github.com/symphonyoss/clabot-config (private) project.

For testing purposes, I've created [https://github.com/symphonyoss/clabot-test/pull/1](https://github.com/symphonyoss/clabot-test/pull/1).

Since the list of contributors is currently empty, I'd have expected the cla-bot to post a comment in the issue, but nothing happened.

Below is reported the `.clabot` contained in `https://github.com/symphonyoss/clabot-config`:

{
  "contributorListGithubUrl": "https://api.github.com/repos/symphonyoss/clabot-config/contents/contributors.json",
  "label": "cla-signed",
  "message": "Thank you for your pull request and welcome to our community! ...."
}

The content of `contributors.json` is `[]`.

**Comment by ColinEberhardt** [comment]
Just grabbing the logs now, so that I can look at it later:

```
START RequestId: 69f43143-4c5d-11e7-ad00-3baf5d20fb3f Version: $LATEST
2017-06-08T15:16:16.451Z	69f43143-4c5d-11e7-ad00-3baf5d20fb3f	callback null { message: 'ignored action of type created' }
END RequestId: 69f43143-4c5d-11e7-ad00-3baf5d20fb3f
REPORT RequestId: 69f43143-4c5d-11e7-ad00-3baf5d20fb3f	Duration: 42.35 ms	Billed Duration: 100 ms Memory Size: 128 MB	Max Memory Used: 36 MB	
START RequestId: d90174ad-4c5d-11e7-827d-876229f0950c Version: $LATEST
2017-06-08T15:19:21.571Z	d90174ad-4c5d-11e7-827d-876229f0950c	Checking CLAs for PR https://api.github.com/repos/symphonyoss/clabot-test/pulls/1
2017-06-08T15:19:21.673Z	d90174ad-4c5d-11e7-827d-876229f0950c	API Request https://api.github.com/repos/symphonyoss/clabot-config/contents/.clabot
{}

2017-06-08T15:19:22.292Z	d90174ad-4c5d-11e7-827d-876229f0950c	Resolving .clabot at project level
2017-06-08T15:19:22.311Z	d90174ad-4c5d-11e7-827d-876229f0950c	API Request https://api.github.com/repos/symphonyoss/clabot-test/contents/.clabot
{}

2017-06-08T15:19:22.490Z	d90174ad-4c5d-11e7-827d-876229f0950c	callback Error: API request https://api.github.com/repos/symphonyoss/clabot-test/contents/.clabot failed with status 404 undefined
2017-06-08T15:19:22.491Z	d90174ad-4c5d-11e7-827d-876229f0950c
{
    "errorMessage": "Error: API request https://api.github.com/repos/symphonyoss/clabot-test/contents/.clabot failed with status 404"
}

END RequestId: d90174ad-4c5d-11e7-827d-876229f0950c
REPORT RequestId: d90174ad-4c5d-11e7-827d-876229f0950c	Duration: 985.33 ms	Billed Duration: 1000 ms Memory Size: 128 MB	Max Memory Used: 42 MB	
```

**Comment by ColinEberhardt** [comment]
From the logs, it first tries to obtain the org level config file:

```
 https://api.github.com/repos/symphonyoss/clabot-config/contents/.clabot
```

But after this it attempts to fetch the project-level file, which implies the above failed. I wonder if the GitHub App requires different access privileges in order to obtain this file?

**Comment by maoo** [comment]
good point! Looking at the (screenshot attached) cla-bot configuration, the red cross close to "No access to code" looks suspicious; would be interesting to try enabling this privilege.

<img width="804" alt="screen shot 2017-06-09 at 18 32 49" src="https://user-images.githubusercontent.com/327285/26985095-4f548554-4d42-11e7-9bbd-233291ebdd07.png">

**Comment by ColinEberhardt** [comment]
That's a very good spot! That would certainly explain a 404 response on a private repo. I've changed the permissions to request read-access for repo data, which should hopefully be suitable.

Once you've accepted this updated permission, let me know and I'll re-run the webhook.

**Comment by maoo** [comment]
Done, thank you!

**Comment by ColinEberhardt** [comment]
That didn't fix it, but I've spotted the problem!

The code that obtains the `.clabot` file is executed before the integration obtains the installation token:

https://github.com/ColinEberhardt/cla-bot/blob/develop/index.js#L47

As a result, these requests are being executed as if from an anonymous user, so do not have the required privileges.

**Comment by maoo** [comment]
Thanks @ColinEberhardt , great catch; I'm wondering if a simple fix like https://github.com/maoo/cla-bot/commit/59e70a57c324ef67ab78185adaf19dbbb747eccd could solve it. wdyt?

**Comment by ColinEberhardt** [comment]
@maoo  the problem with your suggested fix is that the integration token needs to be used in the very initial request here: https://github.com/maoo/cla-bot/commit/59e70a57c324ef67ab78185adaf19dbbb747eccd#diff-168726dbe96b3ce427e7fedce31bb0bcR33

I've done a bit more testing and run out of things to try. I have verified that the GitHub App is able to  access private repo data, so I'm stumped!

Any chance you could make me a collaborator on the clabot-config project so that I can see if there is anything specific to that project that might be causing issues?

**Comment by maoo** [comment]
Done, please accept my invite on https://github.com/symphonyoss/clabot-config/invitations
I had to make the project public in order to invite you; this may actually prevent us from reproducing the issue (though it can direct us to the solution).

**Comment by ColinEberhardt** [comment]
Thanks - although simply changing the repo to public has allowed the bot to function correctly. I have tested that the bot can access private repos using the installation token (I tried with a private ScottLogic repo), so there must be something different about this repo or the way the SymphonyOSS org is structured.

The one immediate difference I can see is that you had to make it public in order to add me as an external collaborator. I don't seem to have to do that for Scott Logic repos.

Do you have any idea why that is the case?

**Comment by maoo** [comment]
The reason _was_ related with billing, since we had the maximum amount of members allowed by our current tier; I've now changed the project to be private again (your invitation is still pending on https://github.com/symphonyoss/clabot-config/invitations ), so we can test the specific use case.

At first glance, I don't see any difference from symphonyoss and ScottLogic github orgs ; I've scrolled through the org configurations, but couldn't find anything suspicious, except one configuration...

<img width="832" alt="screen shot 2017-06-15 at 12 31 58" src="https://user-images.githubusercontent.com/327285/27179473-a91d871a-51c6-11e7-84c1-03192a510e38.png">

As you can see, cla-bot is enabled only on one repo, clabot-test; could this mean that the cla-bot is only allowed to access that repo? Now I added also clabot-config in the list, let's see if things change.

**Comment by maoo** [comment]
I've just submitted [another PR](https://github.com/symphonyoss/clabot-test/pull/3) and this time the cla-bot fails the validation (as expected); I still can't see the comment on the PR though.

**Comment by ColinEberhardt** [comment]
Looking at the logs, I can see that you have the following config:

```json
{
    "contributorListGithubUrl": "https://api.github.com/repos/symphonyoss/clabot-config/contents/contributors.json",
    "label": "cla-signed",
    "message": "Thank you ..."
}
```

When the bot tried to fetch `https://api.github.com/repos/symphonyoss/clabot-config/contents/contributors.json` it returns a 404. Unfortunately when I try to accept the invite for the clabot-config repo, GitHub says it is invalid!

There is one problem with using an `https://api.github.com/.../contents/` URL, this doesn't actually returns the content of the file, instead they returns a download URL for the actual file contents. Currently the bot expects this URL to return the JSON contributors array directly.

I am starting to realise that giving users the information they require to debug their bot settings is not going to be an easy task!

**Comment by maoo** [comment]
I fixed the invitation link (sorry about that, the issue was on the github organisation billing configuration) - https://github.com/symphonyoss/clabot-config/invitations

Regarding the use of contributorListGithubUrl, I've used the api.github.com in conjunction with `getFile` , which actually invokes the `downloadUrl` returned by the first call; you can see the code on https://github.com/ColinEberhardt/cla-bot/blob/3c89e589425670b46754c0346415f059a12f092a/contributionVerifier.js#L16

If this works for public clabot-config repos, should work also for private ones right?

Agreed with the debugging, I see 2 directions to improve this:
1. improve documentation and `npm test` power to create PR/issue local payloads and test them locally with lambdaSpec
2. improve the cla-bot configuration screen (screenshot reported in the comment above), by adding some configuration validation and returning errors if - for example - `.clabot` cannot be correctly parsed

**Comment by ColinEberhardt** [comment]
I think I've fixed it :-)

The request to fetch the GitHub URL was not using the installation token:

https://github.com/ColinEberhardt/cla-bot/commit/6903a523ed2453ecb46b0029a509fe93161644ae#diff-c90bcb623d5ff6ad45cd40e9a2658832

With this bot you can run it in two modes ...

1. As a GitHub App, where the first thing it does is obtain an installation token which it uses to make requests on your repos.
2. As a self-hosted app which is invoked via webhooks. In this case it is your responsibility to ensure the provided access token has the required privileges!

Probably needs a bit more testing before closing this issue, but at least it all finally makes sense!

**Comment by maoo** [comment]
Thanks @ColinEberhardt , that's great! I will close the issue and continue with the configuration to enable cla-bot on all foundation projects.
